### ISSUE TITLE
Documentation doesn't outline how to test if Generic type is null or this feature is missing from null safety

### ISSUE THREAD

**Issue by JohnGalt1717** [dart-lang/sdk#44681]
Dart SDK version: 2.12.0-133.2.beta (beta) (Tue Dec 15 09:55:09 2020 +0100) on "windows_x64"

If I have a class that takes a generic like this:

class Field<TSource, TDestination> {
...
}

Without null safety I can do a check in a function like this:

if (TDestination == String) and this returns correctly.

However, with null safety enabled TDestination can be String? which is fine and good.  But I can't find how to do the same equality comparison. This doesn't work:

if (TDestination == String?) throws an analyzer error.

And I can't find any in the null safety docs or in the dart type system docs either. Neither even makes mention of the first syntax without null safety.

On presumes there is a way to do Type equality with nullables?  If so, how and I believe the docs need to be updated. If not, then I would suggest this needs to be added and my above syntax would make sense given how it currently works without null safety.

Thanks.

**Comment by devoncarew** [comment]
Thanks for the feedback!

Wrt:

> if (TDestination == String?) throws an analyzer error.

Can you post a small self contained snippet, and the analysis error that you see? Thanks!

**Comment by JohnGalt1717** [comment]
Here's an example

```
class Field<TSource, TDestination> {
  void convert(TSource value) {
    if (TSource == TDestination || TSource == TDestination?) {

    }

    if (TDestination == String?) {
      
    }
  }
}
```

The error is "Expected to find ':'.
Expected an identifer.

(on both examples above)

I would have expected it to work as you'd expect because Field<> can be defined with nullables so you need to be able to check it.

I was also expecting that I could limit my generics to not be nullable or allow nullable too, which it appears I cannot?

**Comment by devoncarew** [comment]
OK, I suspect that this is working as intended; I'll cc people on the language team as they'll be able to say why (and, I don't know why the asymmetry around allowing `T == String` but not `T == String?`).

cc @leafpetersen @lrhn

**Comment by lrhn** [comment]
This is working as intended. There is no *syntax* for expressing a nullable type as a type literal expression.

You can write `String` as an expression because you can write an *identifier* as an expression, and that identifier happens to denote a type instead of a variable. There is no ``<identifier> `?'?`` syntax in the Dart expression grammar, so you can't write `String?` as an expression (nor can you write `List<int>` or `int Function(int)`).

What you *can* do to get the `Type` of (almost) any type, is to have a helper function:
```dart
Type typeOf<T>() => T;
```
Then you can write `if (T == typeOf<String?>()) ...`.

That said, I want to *really* discourage any attempt to compare `Type` objects if it can be avoided. It's simply not a very good tool in an object oriented language, and Dart is only really designed to treat type arguments properly generically (so the code works *the same* for any valid type argument). It doesn't work very well for type specialization, where the function works differently if you pass `String` as one of the type arguments. It's doable, but not particularly well supported.
(Consider having as subclass, `class StringField<TSource> extends Field<String, TSource> { ... }` where you can override the methods to be properly specialized).
Also, detecting that `TDestination` is precisely `String` or `String?` only matters if you can be absolutely sure that you know all the relevant types. Here `String` is easy because it has no subtype other than `Never`, but are you absolutely certain that detecting `String` and `String?` is enough, or should you also detect `FutureOr<String>`? What about `FutureOr<FutureOr<String?>?>?`?

**Comment by JohnGalt1717** [comment]
@lrhn This Field class that I'm referring is a generic that takes an input control's value and converts it to a proper value.  There would literally be dozens of Field types that would have to be created as a result of specializing this because it isn't just String => String, or String => String? It's String => int, String => int? etc. etc. through all of the various types that the TextInputField in Flutter supports and event DateTime? => DateTime and DateTime? => DateTime? etc. for pickers etc.

Hence the conversion function uses the TSource and TDestination and the permutations and combinations available to properly handle the conversions.

In my case FutureOr is irrelevent, but absolutely I would expect that I could restrict my generics explicitly to not have what you're suggesting (i.e. I should be able to define that the generic is a primitive type as an example)

And while Type typeOf<T>() => T works fine in some cases there are others where it doesn't.  I would suggest that Dart needs to be expanded to allow T == Type? notation to be able to do this type check natively, and like pretty much all other languages that support generics allow nullability restrictions on the Generic definitions.

Since none of this requires reflection I really think that this is necessary in the dart language. The combination of better generic definition restrictors and being able to test type equality with nullability is required for better code that does what it intends and can't be misused by someone that can't see what the code is doing. (i.e. packages being used by 3rd parties, junior devs that mean well but misuse it, etc)

**Comment by lrhn** [comment]
The `typeOf` function works for every *type*, but it won't be able to make a constant value.
There is currently no way to create a constant `Type` object for a type which cannot be denoted by a single identifier.

When we get generalized type aliases, that will be every type (because you can do `typedef _myIdentifier = anyType;` and refer to it as `_myIdentifier`). 

I'm not sure we can just allow `String?` as expression syntax. For example: `{ x ? (y) : z }` would then be ambiguous if `x` is a type. We need to parse the expression before we can start worrying about what it means, so ambiguity is bad. (On the other hand, we already handle `{x is y ? (z) : w}`, so we can probably do something similar).

**Comment by JohnGalt1717** [comment]
@lrhn  is not TSource == TDestination? completely different in the parser?  The == should denote that there is a test next and if you were doing ? : syntax you'd have to wrap it in () wouldn't you anyhow? Otherwise, it would mean something different would it not? (i.e. TSource == TDestination ? something : somethingElse would have a completely different meaning than TSource == (TDestination ? something : somethingElse) so TSource == TDestination? would parse correctly both because there's no space and because there is no : after it and no brackets.

What about getting limiting generics to primitives and nullable or not nullable?

**Comment by lrhn** [comment]
If you write `(TSource == TDestination?)` then there is currently no valid interpretation for that, so if we allowed `identifier?` type literals, it would be the only valid interpretation. That's not the problem. The problem comes when there is something after the `?`, because then it's possible to write token sequences which can be parsed in two different ways. We can't use types to tell us which one is correct because we can't really talk about any semantic part of the program until after we have successfully parsed the grammar.
That's why I used `{x ? (y) : z}` as example. That can be parsed as `{ (x?)(y) : (z) }`, a map literal where the key is a function call on the `x?` type (which might even be valid if someone declares an extension `call` method on `Type`!), or as `{ (x) ? (y) : (z) }`, a set literal containing a conditional expression.
We need to be able to parse that in one way or another. That complication comes because `?` already means something, so reusing it at an arbitrary point in the language grammar risks parser ambiguities.

**Comment by JohnGalt1717** [comment]
Doesn't the space matter? (Just asking?)

And how does C# and Java make it work? (i.e. why not just copy what they're doing since Dart is so close)

**Comment by lrhn** [comment]
For current Dart parsing, no, spaces do not matter.

Neither C# nor Java have type literals like Dart's. 

You can't write `Integer` as an expression in Java (you use `Integer.class` which is a language feature, not a member access, or `someInteger.getClass()` similar to Dart's `.runtimeType`). There is no `List<int>.class` in Java because `.class` is a run-time feature and generics are not reified at run-time. Java has no non-nullable/nullable distinction, so there is no `Integer?` type at all.  They have `@NonNull` annotations, but that's not part of the *type* system.

C# does have reified generics of a sort, because it does code specialization, and you can get the type of an object with the [`Object.GetType`](https://docs.microsoft.com/en-us/dotnet/api/system.object.gettype?view=net-5.0) method. They have a[`typeof`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#typeof-operator) macro/compiler-feature for expressing arbitrary types, so you can write `typeof(List<string>)`. I assume that also works for nullable types, at least because `int?` can be written as `Nullable<int>` as well.

So, both of those languages have special syntax for expressing types in-line in programs. Dart's "type literals" are easier to write, but also harder to extend to more type syntax because they're not delimited.

**Comment by JohnGalt1717** [comment]
@lrhn  Sounds like an excellent design meeting discussion for the dart team to me... This needs better direction (my vote is make the space matter so it reads naturally like any other type)
